### PR TITLE
Make string_not_contains_constraintt a struct

### PR DESCRIPTION
**Pull request by romainbrenguier**
In particular it does not inherit from exprt anymore.
Not inheriting from exprt, allows the class to be more structured.
We have fields that have named describing what they contain, instead of
having to define accessor methods.
This also prevent us from mistakenly passing a constraint to an exprt
function, which wouldn't know how to deal with a
string_not_contains_constraintt.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 589dd9a73482ad8834e83f4cc835deb7465d376f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90066851

File: src/solvers/refinement/string_constraint.h [hunk 1]
```diff
@@ -116,98 +116,46 @@ inline std::string to_string(const string_constraintt &expr)
 }
 
 /// Constraints to encode non containement of strings.
-class string_not_contains_constraintt : public exprt
+/// string_not contains_constraintt are formula of the form:
```
Comment (majakusber) [inline src/solvers/refinement/string_constraint.h]: :pick: Typo: are formula -> are formulas (or formulae?)
Also typo in commit message: fields that have named -> fields that have names

File: src/solvers/refinement/string_constraint.cpp [hunk 2]
```diff
@@ -46,3 +46,43 @@ string_constraintt::string_constraintt(
     "String constraints must have non-negative upper bound.\n" +
       upper_bound.pretty());
 }
+
+/// Used for debug printing.
+/// \param [in] expr: constraint to render
```
Comment (majakusber) [inline src/solvers/refinement/string_constraint.cpp:51]: Is the `[in]` needed?

**Review by majakusber — APPROVED** [review]
Just minor things, otherwise looks good.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a3318cd4a5405a4a849c249a4ca7a7064a609eea).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90278315

File: jbmc/unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp [hunk 3]
```diff
@@ -282,26 +282,23 @@ SCENARIO("instantiate_not_contains",
...
 
     // Create witness for axiom
     symbol_tablet symtab;
     const namespacet empty_ns(symtab);
     string_constraint_generatort generator(ns);
-    std::map<string_not_contains_constraintt, symbol_exprt> witnesses;
+    std::unordered_map<string_not_contains_constraintt, symbol_exprt> witnesses;
     witnesses[vacuous] = generator.fresh_symbol("w", t.witness_type());
 
     INFO("Original axiom:\n");
```
Comment (peterschrammel) [inline jbmc/unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp:300]: Side remark (not for this PR): Calling debug output INFO is a bit confusing.

**Review by martin-cs — COMMENTED** [review]
I don't see anything intrinsically wrong with this patch but the general approach does not seem quite right to me.  Maybe there are things that I don't see here.  Personally I would lean towards having the expression as an exprt and the information about it that is used during solving as a class / struct.  That way we distinguish between what you are reasoning about and the working information about it.

File: src/solvers/refinement/string_constraint.h [hunk 4]
```diff
@@ -116,98 +116,46 @@ inline std::string to_string(const string_constraintt &expr)
 }
 
 /// Constraints to encode non containement of strings.
-class string_not_contains_constraintt : public exprt
+/// string_not contains_constraintt are formulas of the form:
+/// forall x in [lb,ub[. p(x) => exists y in [lb,ub[. s0[x+y] != s1[y]
```
Comment (peterschrammel) [inline src/solvers/refinement/string_constraint.h]: Can this be explained in terms of the members of the struct?

**Review by peterschrammel — APPROVED** [review]

**Comment by romainbrenguier** [comment]
@martin-cs  
> I don't see anything intrinsically wrong with this patch but the general approach does not seem quite right to me. Maybe there are things that I don't see here. Personally I would lean towards having the expression as an exprt and the information about it that is used during solving as a class / struct. That way we distinguish between what you are reasoning about and the working information about it.

The `string_not_contains_constraintt` is the information that is only used by the solver. The corresponding exprt that appears in the (goto) program are function_application_exprts (for instance "cprover_string_index_of_func")

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 14057f508ef20271befbf920299c2dbe3903495c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90384977

**Comment by martin-cs** [comment]
> The `string_not_contains_constraintt` is the information that is only
> used by the solver. The corresponding exprt that appears in the
> (goto) program are function_application_exprts (for instance
> "cprover_string_index_of_func")

Personally, I'd swap out the function application for a custom exprt,
or have the model implementation call something that resolved into a
custom exprt but it is up to you.

**Pull request merged**